### ISSUE TITLE
Lose the ability to differentiate nil and NSNull null

### ISSUE THREAD

**Issue by michaeleiselsc** [guaranatech/GRNPFObject#1]
why doesn't it just use removeObjectForKey: instead of setting to NSNull? otherwise, you lose the ability to differentiate between the two, but parse still treats them differently. for queries, you could do a similar change

**Comment by anas-ambri** [comment]
Yes, I totally agree with you that the current solution is not ideal, although it is one that worked for us. And I am looking for ways to make it better.

Say, for example, you have the following code:

```
PFQuery *query = [PFComment query];
[query whereKey:@"post" equalTo:post];
[query findInBackground];
```

`post == nil` could be caused by 2 main reasons:
1- Totally a mistake. As in, that shouldn't be `nil` at all. In that case, the best thing to do would probably be not to fail the query.
2- Was genuine, and was done to retrieve a comment for which `post` was actually `undefined`. This is equivalent to replacing `nil` by `NSNull`.

The only way I can see both of these use cases maintained is by providing with a mechanism for annotating a property that could be undefined. Something like this:

```
@interface PFComment: GRNPFObject<Subclassing>
@property (retain) PFPost<GRNPFNonNull> *post;
@end
```

In this case, a comment must always be associated with a post. When we query for comments, and we include the line `[query whereKey:@"post" equalTo:post];`, we expect `post` to always be there. If it is not, then the query should fail.

For case #2, we would have:

```
@interface PFComment: GRNPFObject<Subclassing>
@property (retain) PFPost<GRNPFNullable> *post;
@end
```

In this case, we fall back to the current behaviour.

Does that make sense to you? Would love to hear your feedback!
